File: R/glmIAL.R
glmIAL <-
function(family, y,  X, delta=seq(0, 5.0, by=0.5),  
         tau = c(0.01*c(10,8,6,4,2,1), 0.0001*c(50,20,10,5,2,1)),
         pMax=20, offset=NULL, naPercent=0.4, 
         nTotal=NULL, maxit=50, maxitIAL=20, nReEstimate=pMax, 
         conv=1e-5, fitted=NULL, trace=1, prior=NULL, prop=NULL, 
protect=0)
{
  if(!is.numeric(y)){
    stop("y must be a numeric vector\n")
  }
  
  if(!is.matrix(X)){
    stop("X must be a vector\n")
  }

  if(is.null(prior)) prior=rep(1,length(y))
  
  if(is.null(prop)) prop=1

  M = ncol(X)
  N = length(y)
  
  if(nrow(X) != N){
    stop("the dimension of X and y do not match\n")
  }
  
  if(!is.null(offset)){
    useOffset = 1
    if((! is.numeric(offset)) || length(offset) != N){
      stop("offset must be a numeric vector of the same length as y\n")
    }
  }else{
    useOffset = 0
    offset    = 0.0 
  }
  
  if(family == "binomial"){
    if((! is.numeric(nTotal)) || length(nTotal) != N){
      strWarn = "For binomial family, nTotal must be a numeric vector"
      stop(strWarn, "of the same length as y\n")
    }
  }
  
  if(is.null(fitted)){
    init = 0
    fitted = rep(0, N)
  }else{
    init = 1
    if((! is.numeric(fitted)) || length(fitted) != N){
      stop("fitted must be a numeric vector of the same length as y\n")
    }
  }
  
  #define BINOMIAL  1
  #define POISSON   2
  #define GAMMA     4
  
  #define LOGIT     1
  #define LOG       2
  #define INVERSE   4
  
  if(family=="binomial"){
    familyR = 1
    linkR   = 1
  }else if(family=="poisson"){
    familyR = 2
    linkR   = 2
  }else if(family=="gamma"){
    familyR = 4
    linkR   = 4
  }else if(family=="gaussian"){
    familyR = 3
    linkR   = 3
  }else{
    stop("invalid family\n")
  }
  
  isNA = apply(X, 1, function(v){ any(is.na(v)) })
  isNA = is.na(y) | isNA
  
  if(length(which(isNA))/N > naPercent){
    stop("percent of missing data is too high\n")
  }
  
  w2kp  = which(!isNA)
  yR    = y[w2kp]
  XR    = X[w2kp,]
  wXR   = XR
  nXR   = matrix(0, nrow=N, ncol=max(nReEstimate+1,1))
  
  if(useOffset){
    offset = offset[w2kp]
  }
  if(family == "binomial"){
    nTotal = nTotal[w2kp]
  }

  N  = length(yR)
  
  protect=protect-1

  dims    = numeric(11)
  dims[1] = N
  dims[2] = M
  dims[3] = maxit
  dims[4] = maxitIAL
  dims[5] = init
  dims[6] = useOffset
  dims[7] = length(delta)
  dims[8] = length(tau)
  dims[9] = pMax
  dims[10] = nReEstimate
  dims[11]= protect
  
  resid   = weights = numeric(N)
  fitted  = fitted[w2kp]
  nIter   = 0
  phi     = 0.0
  
  w2use   = rep(-9, pMax)
  b2use   = numeric(pMax)
  score   = matrix(0, nrow=length(delta), ncol=length(tau))
  b0      = matrix(0, nrow=length(delta), ncol=length(tau))
  b02use  = 0.0
  convg   = 0
  n2use   = 0
  score2use = 0.0
  delta2use = 0.0
  tau2use   = 0.0
  wss  = numeric(maxit)
  wss0 = 0
  
  Z = .C("glmIAL", as.integer(familyR), as.integer(linkR), 
         as.integer(dims), nIter=as.integer(nIter), 
         as.double(yR), as.double(offset), as.double(XR), 
         as.double(wXR), as.double(nXR), as.double(nTotal), 
         as.double(conv), fitted=as.double(fitted), as.double(resid), 
         as.double(weights), as.double(phi), as.integer(trace), 
         n2use=as.integer(n2use), w2use=as.integer(w2use), 
         b2use=as.double(b2use), b02use=as.double(b02use),
         b0=as.double(b0), score=as.double(score), 
         score2use=as.double(score2use), as.double(delta), 
         delta2use=as.double(delta2use), as.double(tau), 
         tau2use=as.double(tau2use), wss=as.double(wss),
         wss0=as.double(wss0), as.double(prior), as.double(prop),PACKAGE="hmmcov")
  
  n2use=Z$n2use
  
  if(n2use > 0){
    w2use = Z$w2use[1:n2use]+1
    b2use = Z$b2use[1:n2use]
  }else{
    w2use = b2use = NULL
  }
  
  if (Z$nIter >= maxit){
    warning(sprintf("reach mamximum %d iterations\n", maxit))  
  }
  
  score  = matrix(Z$score, nrow=length(delta), ncol=length(tau), byrow=TRUE)
  b0     = matrix(Z$b0,    nrow=length(delta), ncol=length(tau), byrow=TRUE)
  wss    = Z$wss[1:Z$nIter]

  result = list(n2use=n2use, w2use=w2use, b2use=b2use, b0=b0, 
                score2use=Z$score2use, delta2use=Z$delta2use, 
                tau2use=Z$tau2use, score=score, nIter=Z$nIter, 
                wss=wss, wss0=Z$wss0, fitted=Z$fitted)
  
  result
}

File: src/glmIAL.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <string.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/Applic.h>
#include "utility.h"
#include "glm.h"
#include "IAL.h"

/**********************************************************************
 *
 * glmIAL
 *

 
 Input:
 
 family       GLM family (see below)
 link         Link function (see below)
 N            # units
 M            # X variables
 y            y-variable (N-vector)
 X            If M>0, N*M matrix of X variables
 maxit        Maximum number of iterations of IRLS algorithm
 conv         Proportional change in weighted sum of squares residuals to
              declare convergence
 init         If true (non-zero), the iteration starts from initial estimates 
              of fitted values (see below). This option has no effect if
              no iteration is required
 
 Output:
 
 fitted       fitted values 
 resid        working residuals (on linear predictor scale) (N-vector)
 weights      weights (N-vector)
 beta         regression coeficients

 Return
 
 0            convergence
 1            no convergence after maxit iterations
 
 **********************************************************************/

int glmIAL(int* familyR, int *linkR, int *dims, int *nIter,
           double *y, double *offset, double *RX, 
           double *RwX, double *RnX, double *nTotal_binom,  
           double *convR, double *fitted, double *resid,  
           double *weights, double *phi, int *trace, int *n2use,  
           int *w2use, double *b2use, double* b02use, double* Rb0M, 
           double *Rscore, double *score2use, double *delta, 
           double *delta2use, double *tau, double *tau2use, 
           double *wssV2, double *wss0,
           double *prior, double *prop)
{
  double epsilon = 1e-8;       /* Singularity threshold */
  int N, M, maxit, maxitIAL, init, useOffset, n_delta, n_tau;
  int i, j, k, k1, k2, Nu, p_max, n2kpIAL=0, *w2kpIAL;
  int convged = 0, iter = 0, useInitIAL = 0, nReEstimate;
  
  int family = *familyR;
  int link   = *linkR;
  
  if(family > 6 || family < 1){
    Rprintf("family=%d, ", family);
    error("Invalid family!\n");
  }
  
  double conv = *convR;
  double mu, eta, ri, wi, wsum, D, Vmu, wss, wss_last,pi;
  double **score, *z, *wz, xij, **X, **wX, **nX, **b0M;    
  double *b, *v, *Xj2, *Xj, *wXj, BIC=0.0, *resIAL;
  double delta1, tau1, *b2kpIAL, b0=0.0, Xjnorm, *wssV1,penalty=0;
  
  *score2use = DBL_MAX;

  N         = dims[0];
  M         = dims[1];
  maxit     = dims[2];
  maxitIAL  = dims[3];
  init      = dims[4];
  useOffset = dims[5];
  n_delta   = dims[6];
  n_tau     = dims[7];
  p_max     = dims[8];
  nReEstimate = dims[9];
  
  /* Note, each row of X and wX is one covariate */
  reorg(RX,  &X,  M, N);
  reorg(RwX, &wX, M, N);
  k = nReEstimate+1;
  if(k < 1) { k = 1; }
  reorg(RnX, &nX, k, N);  
  reorg(Rscore, &score, n_delta, n_tau);
  reorg(Rb0M,   &b0M,   n_delta, n_tau);

  z  = (double *)Calloc(N, double);  // working y
  wz = (double *)Calloc(N, double);  // weighted working y
  /* variables to be send to function IAL */
  b  = (double *)Calloc(M, double);
  v  = (double *)Calloc(M, double);
  
  wssV1   = (double *)Calloc(maxit, double);
  resIAL  = (double *)Calloc(N, double);
  w2kpIAL = (int *)Calloc(M, int);
  b2kpIAL = (double *)Calloc(M, double);

  /**
   * sum square for each marker, i.e., each column of X
   */
  Xj2 = (double *)Calloc(M, double);
  
  if(*trace > 7){
    Rprintf("y\n");
    Rprint_ve(y, 0, 9);
    Rprintf("X\n");
    Rprint_me(X, 0, 1, 0, 9);
  }
  
  if(*trace>3)
    Rprintf("  glmIAL: family=%d, phi=%f, init=%d\n", family, *phi, init);

  /* ----------------------------------------------------------------
   * lopp across different deltas and taus
   * ----------------------------------------------------------------*/
  
  for(k1=0; k1 < n_delta; k1++){
    
    delta1 = delta[k1];
    
    if(*trace > 5){
      Rprintf("\nk1=%d, delta1=%f\n", k1, delta1);
    }
    
    for(k2=0; k2 < n_tau; k2++){
      
      tau1 = tau[k2];
      
      if(*trace > 5){
        Rprintf("\n  k2=%d, tau1=%f\n", k2, tau1);
      }
      
      /* ----------------------------------------------------------------*
       * by default, initialize mu (fitted) by y itself, with neccesary 
       * modification, e.g., y + 0.1 to avoid log(0) for Poisson family
       * ----------------------------------------------------------------*/
      if (!init) {
        initialize(family, y, fitted, N, nTotal_binom);
      }
      
      /* ----------------------------------------------------------------*
       * Initialize wi (weights) and (standardized) residual 
       * In IRLS, z_i = eta_i + (y_i - mu_i)*(d_eta/d_mu)_i
       *
       * eta_i = linkfun(mu_i)
       * mu_i  = invlink(eta_i)
       *
       * In the following code:
       * ri = (y_i - mu_i)*(d_eta/d_mu)_i
       * wi = Vmu is the weight,  
       *
       * for those invlaid values, set their weight to be 0.
       * ----------------------------------------------------------------*/
      
      wsum = 0.0;
      wss  = 0.0;
      
      for (i=0; i<N; i++) {
    	
	pi=prior[i];
        mu = fitted[i];
        
        if (!muvalid(family, mu)) { 
          wi = ri = 0.0; 
        }else {
          Vmu = varfun(family, mu, *phi);
          
          if (link == family) {
            ri = (y[i] - mu)/Vmu;
            wi = pi*Vmu;
          }else {
            D  = dlink(link, mu);
            ri = D*(y[i] - mu);
            wi = pi/(D*D*Vmu);
          }
          
          wss += wi*ri*ri;
        }
        
        weights[i] = wi;
        resid[i]   = ri;
        if (weights[i] < epsilon) weights[i] = 0.0;
        
        wsum += weights[i];
      }
      
      /* this is the best wss0 we can get if fitted value is initialized by y */
      *wss0 = wss;

      /* ----------------------------------------------------------------*
       * If summation of all weights is too small, stop 
       * ----------------------------------------------------------------*/
      
      if (wsum < epsilon) {
        Rprintf("  glmIAL: summation of all weights are too small!\n");
      }
      
      if(*trace > 5){
        Rprintf("\n  glmIAL: finish initialization, N=%d, M=%d, family=%d\n", 
                N, M, family);
      }
      
      /* ----------------------------------------------------------------*
       * IRLS algorithm  
       * ----------------------------------------------------------------*/
      
      convged  = 0;
      iter     = 0;
      wss_last = 0.0;

      while(iter<maxit && !convged) {
        
        if (*trace > 5) {
          Rprintf("\n  glmIAL: iteration %d: \n", iter);
        }
                
        for (i=0; i<N; i++) {
          /**
           * current estimate of eta + (y-mu)/gradient
           *
           * linkfun(link, fitted[i]) = eta
           * resid[i] = (y-mu)/gradient
           */
          z[i] = linkfun(link, fitted[i]) + resid[i];
        }
        
        if (useOffset) {
          for (i=0; i<N; i++) z[i] -= offset[i];
        }
        
        /**
         * incoporate the weights to z 
         */
        
        for (i=0; i<N; i++) {
          wz[i] = sqrt(weights[i])*z[i];
        }
        
        /**
         * incoporate the weights to X
         * and calculate the scale of X
         */
        
        for (j=0; j<M; j++) {
          Xj     = X[j];
          wXj    = wX[j];
          Xjnorm = 0.0;
          
          for (i=0; i<N; i++) {
            xij     = sqrt(weights[i])*Xj[i];
            wXj[i]  = xij;
            Xjnorm += xij*xij;
          }
          
          Xj2[j] = Xjnorm;
        }
        
        if (iter==0) {
          useInitIAL=0;
        }else {
          useInitIAL=1;
        }

        IAL(wz, wX, nX, &b0, b, v, delta1, tau1, dims, Xj2, conv, resIAL,  
            &BIC, &n2kpIAL, w2kpIAL, b2kpIAL, useInitIAL, *trace, nReEstimate, weights,prop);
        
//        if (n2kpIAL == 0) { break; }
        
        /**
         * update the fitted values and weights  
         */
        
        wss = 0.0;
        Nu  = 0;
        
        for (i=0; i<N; i++) {
          
          eta = b0;
          for (k=0; k<n2kpIAL; k++) {
            j    = w2kpIAL[k];
            eta += X[j][i]*b2kpIAL[k];
          }
           
          if (useOffset) { eta += offset[i]; }
          
          pi=prior[i];
          mu = invlink(link, eta);
          
          fitted[i] = mu;
          
          if (weights[i]<=0.0) {
            wi = ri = 0.0;
          } else {
            
            Vmu = varfun(family, mu, *phi);
            Nu ++;
            
            if (link == family) {
              ri = (y[i] - mu)/Vmu;
              wi = pi*Vmu;
            }else {
              D  = dlink(link, mu);
              ri = D*(y[i] - mu);
              wi = pi/(D*D*Vmu);
            }
            if (wi < epsilon) wi = 0.0;

            wss += wi*ri*ri;
            
          }
          
          weights[i] = wi;
          resid[i]   = ri;
        }
        
        if(*trace > 9){
          if (n2kpIAL > 0) {
            Rprintf("w2kpIAL: ");
            Rprint_vi(w2kpIAL, 0, n2kpIAL-1);
            Rprintf("b2kpIAL: ");
            Rprint_ve(b2kpIAL, 0, n2kpIAL-1);        
          }
        }
        
        if(wss > 1/epsilon){
          if(*trace > 5){
            Rprintf("  glmIAL: huge wss (%,3e)!\n", wss);
          }
          
//          break;
        }
        
        if (*trace > 5) {
          Rprintf("    n2kpIAL=%d, wss=%.3e\n", n2kpIAL, wss);
        }
        
        wssV1[iter] = wss;
        convged  = (Nu<=0) || (fabs(wss-wss_last)/(wss_last + 0.1) < conv);
        wss_last = wss;
        iter++;
                
      }
      
      if(*trace > 5){
        Rprintf("    n2kpIAL=%d, BIC=%.3e\n", n2kpIAL, BIC);
      }
      
      if(convged==0 /*|| n2kpIAL > p_max*/ & *trace>5){	
	Rprintf("glmIAL: Failure to Converge\n");
      }

      //Hopefull this will not result in an error, but want to report even if just intercept
      if(n2kpIAL==0){
//        continue;
      }
      
      if(*trace > 5){
        if (n2kpIAL > 0) {
          Rprintf("w2kpIAL: ");
          Rprint_vi(w2kpIAL, 0, n2kpIAL-1);
          Rprintf("b2kpIAL: ");
          Rprint_ve(b2kpIAL, 0, n2kpIAL-1);          
        }
      }
      
      score[k1][k2] = BIC;
      b0M[k1][k2]   = b0;

      if (BIC < *score2use) {
        *score2use = BIC;
        *delta2use = delta1;
        *tau2use   = tau1;
        *n2use     = n2kpIAL;
        *b02use    = b0;
        *nIter     = iter;

        for (i=0; i < n2kpIAL; i++) {
          w2use[i] = w2kpIAL[i];
          b2use[i] = b2kpIAL[i];
	  			//penalty += *prop*(1 + *delta2use)*log(fabs(b2use[i]) + *tau2use);
	  			//*score2use = penalty;
        }
        
        for (i=n2kpIAL; i < p_max; i++) {
          w2use[i] = -9;
          b2use[i] = 0.0;
        }
        
        for (i=0 ; i < iter; i++) {
          wssV2[i] = wssV1[i];
        }
        
        if(iter < maxit){
          for (i=iter ; i < maxit; i++) {
            wssV2[i] = 0.0;
          }
        }
        
      }

    }
  }
  
  if(*trace > 5){
    Rprintf("\nn2use = %d, score2use=%.2e\n", *n2use, *score2use);
    Rprintf("score: \n");
    Rprint_me(score, 0, n_delta-1, 0, n_tau-1);

    if (*n2use > 0) {
      Rprintf("w2use: \n");
      Rprint_vi(w2use, 0, *n2use-1);
      Rprintf("b2use: \n");
      Rprint_ve(b2use, 0, *n2use-1);          
    }
  }

  Free(z);
  Free(wz);
  Free(Xj2);
  Free(b);
  Free(v);
  Free(wssV1);
  Free(w2kpIAL);
  Free(b2kpIAL);
  Free(resIAL);

  return(*n2use);
}
